Support for motion detection

miquelruiz opened the pull request:
This adds 3 new methods to the public API:
 - get_motion_detection: returns a boolean telling if motion detection is enabled or disabled
 - enable_motion_detection
 - disable_motion_detection

In addition to that, changes a bit the logging format. The main reason for that change was to add the line number to quickly identify where things were being logged.

Based on the feedback on this I'm preparing another pull request to support enabling/disabling/getting the PIR alarm.

mezz64 commented on pyhik/hikvision.py:
```diff
@@ -132,6 +134,72 @@ def current_event_states(self):
+
+        try:
+            tree = ET.fromstring(response.text)
+            nmsp = tree.tag.split('}')[0].strip('{')
```
You shouldn't need to check the namespace a second time here.  We fetch it during the device info call in the main initialization function.

mezz64 replied:
I see that it does indeed change with the URL path you've choosen, but it remains consistent when using the one defined in the API docs that I had saved as a comment:
http://X.X.X.X/ISAPI/System/Video/inputs/channels/1/motionDetection

Is there a reason why you choose a different endpoint?

miquelruiz replied:
That endpoint returns all the scheduling data together with the state of motion detection. The response on the one I used contains very little additional data, and I thought it was unnecessary to be sending/retrieving all that data if there's no need.

Provided that the XML schema indeed changes between different calls, I consider it more safe to always check it and set it accordingly.

mezz64 commented on pyhik/hikvision.py:
```diff
@@ -132,6 +134,72 @@ def current_event_states(self):
         """Return Event states dictionary"""
         return self.event_states
 
+    def get_motion_detection(self):
```
You can leave this public, but i'd consider the action of checking the current status something that would mostly be called internally.  To adopt as core functionality I think it makes sense to add it to the initialize call so that state is determined right away.  Change it to update a variable, rather than return directly.  A property can be defined to return the variable so I/O isn't done every time we want to see what the state is.

The set function already does an update before making a change so we should always be correct internally.

miquelruiz replied:
I like having it public so it can be integrated with HomeAssistant, for example.
I'll add the variable to the class and update it from here.

Regarding not doing I/O, I'd like a way to force making the call because someone may change this from the camera UI, and then it'll get out of sync.

mezz64 replied:
Integration with HASS is the driver behind most of my suggestions. :) We want to keep I/O out of the event loop.  Saving the state to a variable and defining a property will prevent that.

miquelruiz commented:
I've added the "motion_detection" property that can be used to read the state without making any network call, and left the "get_motion_detection" method public to force the call to happen.

Hope I got it right that's what you meant :)

I've also deleted the logging config call since it should be up to the caller to setup logging however they want.

mezz64 commented on pyhik/hikvision.py:
```diff
@@ -132,6 +132,77 @@ def current_event_states(self):
         """Return Event states dictionary"""
         return self.event_states
 
```
Please add a property here to return the current motion detection state and remain consistent.

```suggestion
@property
def current_motion_detection_state(self):
    """Return current state of motion detection property"""
    return self.motion_detection
```

mezz64 commented on pyhik/hikvision.py:
```diff
@@ -132,6 +132,77 @@ def current_event_states(self):
+
+    def _set_motion_detection(self, enable):
+        url = '%s/MotionDetection/1' % self.root_url
+        self.get_motion_detection()
```
What happens if we send a request to enable motion detection and it's already turned on?  Is it possible we can eliminate this call and rely on a 200 response from the camera to ensure we get the state we want?

mezz64 commented on pyhik/hikvision.py:
```diff
@@ -186,6 +257,8 @@ def initialize(self):
         else:
             _LOGGING.debug('No Events available in dictionary.')
 
+        self.motion_detection = self.get_motion_detection()
```
No need to assign the return to the variable since the function does it on it's own.  Just `self.get_motion_detection()` should be fine.

mezz64 commented on pyhik/hikvision.py:
```diff
@@ -132,6 +132,84 @@ def current_event_states(self):
+    def _set_motion_detection(self, enable):
+        url = '%s/MotionDetection/1' % self.root_url
+        current_state = self.get_motion_detection()
+        if enable == current_state
```
Invalid syntax. Please test any changes before submitting.

This also isn't necessary in the tests I ran.  The camera doesn't care if we send a request for the state it's already in.

miquelruiz replied:
Sure, the camera may not care, but saves an unnecessary network request. I can delete it, though.

mezz64 commented on pyhik/hikvision.py:
```diff
@@ -132,6 +132,84 @@ def current_event_states(self):
+
+    def _set_motion_detection(self, enable):
+        url = '%s/MotionDetection/1' % self.root_url
+        current_state = self.get_motion_detection()
```
Remove.

mezz64 commented on pyhik/hikvision.py:
```diff
@@ -132,6 +132,84 @@ def current_event_states(self):
+        if response.status_code != requests.codes.ok:
+            # If we didn't receive 200, abort
+            _LOGGING.error('Unable to set motion detection: %s', response.text)
+
```
If we get no errors from the camera we should update `self.motion_detection` with the enable value.

miquelruiz replied:
Good catch! I'll add a few unit tests.
